# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/generic/LocalAccessible.cpp
# Commit: b92893ebb007
# Full Hash: b92893ebb007f76bc8b7a9c070e2938a0598a381
# Author: James Teh <jteh@mozilla.com>
# Date: 2021-10-08 09:48:33
# Regressor Bug: 1734322
# File Overlap Count: 1
# Description:
#   Bug 1734322: Cash DOM node id in the parent process a11y cache. r=eeejay
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D127757
# ==============================================================================

diff -r 92bfb5c41eb7 -r b92893ebb007 accessible/generic/LocalAccessible.cpp
--- a/accessible/generic/LocalAccessible.cpp	Fri Oct 08 00:52:39 2021 +0000
+++ b/accessible/generic/LocalAccessible.cpp	Fri Oct 08 01:57:46 2021 +0000
@@ -3066,7 +3066,13 @@
   }
 
   DocAccessibleChild* ipcDoc = mDoc->IPCDoc();
-  MOZ_ASSERT(ipcDoc);
+  if (!ipcDoc) {
+    // This means DocAccessible::DoInitialUpdate hasn't been called yet, which
+    // means the a11y tree hasn't been built yet. Therefore, this should only
+    // be possible if this is a DocAccessible.
+    MOZ_ASSERT(IsDoc(), "Called on a non-DocAccessible but IPCDoc is null");
+    return;
+  }
 
   RefPtr<AccAttributes> fields =
       BundleFieldsForCache(aCacheDomain, aUpdateType);
@@ -3153,6 +3159,16 @@
     }
   }
 
+  if (aCacheDomain & CacheDomain::DOMNodeID) {
+    MOZ_ASSERT(mContent);
+    nsAtom* id = mContent->GetID();
+    if (id) {
+      fields->SetAttribute(nsGkAtoms::id, id);
+    } else if (aUpdateType == CacheUpdateType::Update) {
+      fields->SetAttribute(nsGkAtoms::id, DeleteEntry());
+    }
+  }
+
   return fields.forget();
 }
 